### PR TITLE
Iterate `Worksheet.table` instead of indexing

### PR DESCRIPTION
**Pull request by adriandelgado**
The main change is a refactor of the `write_data_table` and `calculate_spans` functions. This required to change the `Worksheet.table` to use `BTreeMap`s because we need the keys to be ordered.

Some secondary changes were made to the functions inside the main loop of `write_data_table`.

image

Note: the code in `calculate_spans` is still indexing the rows. I wasn't sure if I could safely change it to an iteration.

### CODE REVIEW

File: src/worksheet.rs [hunk 1]
```diff
@@ -7609,17 +7609,18 @@ impl Worksheet {
 
     // Insert a cell value into the worksheet data table structure.
     fn insert_cell(&mut self, row: RowNum, col: ColNum, cell: CellType) {
-        match self.table.get_mut(&row) {
-            Some(columns) => {
+        use std::collections::btree_map::Entry;
```
Comment (jmcnamara) [inline src/worksheet.rs:7612]: Move to the top with other use statements.

File: src/worksheet.rs [hunk 2]
```diff
@@ -7697,14 +7698,10 @@ impl Worksheet {
...
     // Cached/faster version of utility.col_to_name() to use in the inner loop.
-    fn col_to_name(&mut self, col_num: ColNum) -> String {
-        if let Some(col_name) = self.col_names.get(&col_num) {
-            col_name.clone()
-        } else {
-            let col_name = utility::col_to_name(col_num);
-            self.col_names.insert(col_num, col_name.clone());
-            col_name
-        }
+    fn col_to_name(col_names: &mut HashMap<u16, String>, col_num: ColNum) -> &str {
```
Comment (jmcnamara) [inline src/worksheet.rs:7701]: Is there a performance benefit to converting this from a method to an associated function? If not, and this is a semantic change, then I'd prefer it to stay as a method. If it improves performance then the change is okay.

Comment (adriandelgado) [reply src/worksheet.rs:7701]: `col_to_name` is a hot function. Without this change, we would need to clone the `&str` in order to use `&mut self.writer.xmlfile` afterwards. Those clones (allocations) had a measurable difference and showed up in flamegraph.

File: src/worksheet.rs [hunk 3]
```diff
@@ -9111,43 +9093,55 @@ impl Worksheet {
...
     // Write the <c> element for a number.
-    fn write_number_cell(&mut self, row: RowNum, col: ColNum, number: &f64, xf_index: &u32) {
-        let col_name = self.col_to_name(col);
-        let mut style = String::from("");
+    fn write_number_cell(&mut self, row: RowNum, col: ColNum, number: f64, xf_index: u32) {
+        let col_name = Self::col_to_name(&mut self.col_names, col);
 
-        if *xf_index > 0 {
-            style = format!(r#" s="{}""#, *xf_index);
+        if xf_index > 0 {
```
Comment (jmcnamara) [inline src/worksheet.rs:9099]: +1 to this and similar changes below to remove the allocation.

File: src/worksheet.rs [hunk 4]
```diff
@@ -9280,44 +9279,61 @@ impl Worksheet {
...
 
         self.writer.xml_end_tag("cols");
     }
 
     // Write the <col> element.
-    fn write_col(&mut self, first_col: &ColNum, last_col: &ColNum, col_options: &ColOptions) {
-        let mut attributes = vec![];
-        let first_col = *first_col + 1;
-        let last_col = *last_col + 1;
+    fn write_col(
```
Comment (jmcnamara) [inline src/worksheet.rs:9326]: Same method -> associated function comment as above.

**Comment by jmcnamara** [comment]
Overall looks good and not far off merge.

File: src/worksheet.rs (hunk 2 above)
Comment (jmcnamara) [reply src/worksheet.rs:7701]: Ok. +1 to this change in that case.

Comment (jmcnamara) [reply src/worksheet.rs:7701]: One other possible optimisation is that for columns <= 25 we could have a shortcut conversion from the index to the ASCII char like this code from utility.rs:

```rust
let col_letter = char::from_u32(64u32 + remainder as u32).unwrap();
```

Or simpler still a indexed char slice of a str like `"ABCDEF...XYZ"`.

File: src/worksheet.rs (hunk 4 above)
Comment (jmcnamara) [reply src/worksheet.rs:9326]: Yes. I think leave this one as it is. The other one makes more sense.

File: src/worksheet.rs (hunk 2 above)
Comment (adriandelgado) [reply src/worksheet.rs:7701]: Good idea for a future PR

**Comment by jmcnamara** [comment]
Merged. Thanks.

**Pull request merged**